### DISCUSSION TITLE
Is top level await supported?

### DISCUSSION THREAD

**Discussion by aegatlin** [vercel/next.js#11185] [Help] [answered]
Hello all,

When I try to use top-level `await` with Typescript 3.8, and appropriate `tsconfi.json` file, I get a Syntax Error. I can be more thorough in my error logs if it turns out it is supported. But, I have this feeling that there is something in the middleware of Nextjs that would need to be updated before it's supported, but I'm not sure.

So, is top-level `await` supported yet?

Thanks you!

**Comment by timneutkens** [comment]
> So, is top-level await supported yet?

It's not supported currently:
https://babeljs.io/docs/en/babel-plugin-syntax-top-level-await

We use webpack and webpack 5 is not out yet.

**Reply by aegatlin** [reply]
Thank you! The internals of Nextjs feel a bit like a black box to me, so I wasn't sure where to start looking to solve my own problem in this regard. To that end: how would I have been able to find out this information myself?

**Comment by agustif** [comment]
Hey  @timneutkens since we have webpack 5 with next 9.5 now, could this be supported?

https://webpack.js.org/configuration/experiments/

**This code doesnt work!!**
```webpack.config.js

module.exports = {
  //...
  experiments: {
    topLevelAwait: true,
  },
};
```


```next.config.js
module.exports = {
  webpack: (config, options) => {
const {experiments} = options;
config.module.rules.push({
    use: [options.experiments, {options: {topLevelAwait: "true"}}]
})
    return config
  },
}
```
```babelrc
{
  "presets": [
    [
      "next/babel",
      {
        "preset-env": { "targets": {"node": "current"}},
      },
    ]
  ],
  "plugins": [["@babel/plugin-proposal-decorators", { "legacy": true }], ["@babel/plugin-syntax-top-level-await"]],
}
```

**Reply by jensmeindertsma** [reply]
This would be really useful!

**Reply by agustif** [reply]
> This would be really useful!

Hey, I actually got it to work yesterday!!

You just need to be using latest next.js 9.5.x release, and enable webpack5 beta by adding it to your resolutions in your package.json, and modify webpack config via editing next.config.js

```package.json
...
    "resolutions": {
        "webpack": "^5.0.0-beta.28"
    }
...
```
Will upload a minimal functional example later!

but it's just a matter of adding:

```next.config.js
....
config.experiments = {topLevelAwait: true}
return config;
...
```

Hope it helps!!

No need to use babel, a lot of info about next + babel can be outdated like the babel es2017 plugins etc, better to just do it all from package.json and next.config.js (if you don't already have babelrc / babel in your project anyway)

**Reply by jensmeindertsma** [reply]
```
....
config.experiments = {topLevelAwait: true}
return config;
...
```
This goes inside the `webpack` property inside the Next.js config right? Thanks!

**Reply by agustif** [reply]
> ```
> ....
> config.experiments = {topLevelAwait: true}
> return config;
> ...
> ```
> 
> This goes inside the `webpack` property inside the Next.js config right? Thanks!

Yup like this:
```next.config.js
module.exports = {
  webpack: (config, options) => {
    config.experiments = {
        "topLevelAwait": true
    }
    return config
  },
}
```

**Reply by nakazoomi** [reply]
Thank you! It worked :)

**Comment by agcty** [comment]
**EDIT**: It works but you need to use yarn because only yarn understands "resolutions"

Did someone manage to get this working with next 10? I've tried the solution outlined above but it doesn't work:

next.config.js
```javascript
module.exports = {
  webpack: (config, options) => {
    config.experiments = {
      topLevelAwait: true,
    };
    return config;
  },
};

```

package.json

```javascript
"resolutions": {
    "webpack": "^5.0.0-beta.30"
  }
```

but I get this error:

![image](https://user-images.githubusercontent.com/12101091/99047477-8d3e7400-2594-11eb-863c-2d4c6ae99a1b.png)

**Reply by alimertcakar** [reply]
I am experiancing the same. Have you found a solution?

**Reply by alimertcakar** [reply]
Ive got resolutions and all

**Reply by retronav** [reply]
According to this : 
![image](https://user-images.githubusercontent.com/52596591/107619011-7e8d5900-6c78-11eb-8858-0ceb55920fc3.png)
and this(from the code):
https://github.com/vercel/next.js/blob/6b99bda3e3971831c0997ce59b19cb848e4099ff/packages/next/next-server/server/config.ts#L95

You should include 
```js
future: {
    webpack5: true,
  },
```
in your next.config.js and then webpack 5 should work.

**Comment by Janpot** [comment, marked as answer]
In https://github.com/vercel/next.js/pull/17590 I've added support for the experimental `topLevelAwait` option of webpack. To use it, as per https://github.com/vercel/next.js/discussions/11185#discussioncomment-55483 make sure you're on webpack 5 and enable the experimental option.

For an example you can check the integration tests for this: https://github.com/vercel/next.js/tree/canary/test/integration/async-modules

Webpack 5 instructions: https://nextjs.org/blog/next-9-5#webpack-5-support-beta

**Reply by agcty** [reply]
thank you! it's important to note though that you have to use yarn

**Reply by besSejrani** [reply]
Does this still apply with Next 10.1 ? I tried like in the example, but top-level await doesn't work.

**Reply by eric-burel** [reply]
This sounds suddenly broken for me when switching to Next 13/app folder

**Comment by balazsorban44** [comment]
Are there any good/interesting usecases for this feature?

**Reply by Janpot** [reply]
```js
import { shouldPolyfill } from '@formatjs/intl-numberformat/should-polyfill';
if (shouldPolyfill()) {
  await Promise.all([
    import('@formatjs/intl-numberformat/polyfill'),
    import('@formatjs/intl-numberformat/locale-data/en'),
  ]);
}
```
Caveat: this still fails on ie11 (https://github.com/vercel/next.js/pull/18487)

**Reply by balazsorban44** [reply]
Hmm. We are actually in the process of adding polyfills to our app at work. I figured we just use `<script nomodule/>` in `_document`. What's the point of polyfills if they cannot be loaded in IE11? 🤔 (not meaning top level await is wrong, but I wouldn't consider your suggestion a valid usecase). Thanks anyway! Any other ideas? There had to be a good reason to add this feature, right?

**Reply by Janpot** [reply]
> What's the point of polyfills if they cannot be loaded in IE11?

This is an experimental feature, the goal is to make it work on ie11.

**Comment by onceLearner** [comment]
what should I do as a Windows NPM user ??  is wrapping top-level await with async function  is the only solution right now ?

**Comment by karlhorky** [comment]
Seems like Top-Level Await via the `config.experiments.topLevelAwait` configuration key is currently broken in Next.js 12 (version `12.0.1`):

https://github.com/vercel/next.js/pull/30082#issuecomment-952673352

**Comment by sajadghawami** [comment]
i can confirm that. Getting:

```
Error: error: top level await requires target to es2017 or higher and topLevelAwait:true for ecmascript
```

Even though the exact same config worked in `nextjs@11` without a problem.

**Reply by ivan-kleshnin** [reply]
Make it an issue please, the team does not generally track **Discussions** for bugs.

**Comment by teamteak** [comment]
Getting the same issue on our Vercel deployment and never had the same issue before.

`Error: error: top level await requires target to es2017 or higher and topLevelAwait:true for ecmascript
`

**Reply by kazdhicham** [reply]
get the same probleme when adding api server

**Reply by flieks** [reply]
i am also getting this error
```
Module parse failed: The top-level-await experiment is not enabled (set experiments.topLevelAwait: true to enabled it)
File was processed with these loaders:
 * ./node_modules/next/dist/build/webpack/loaders/next-swc-loader.js
```

**Comment by felipe-muner** [comment]


**Reply by sajadghawami** [reply]
what next version are you using?

**Reply by felipe-muner** [reply]
Hello @sajadghawami, 

I was wrong, actually I changed the tsconfig and it worked.
I was using the latest version for that day....

See you

**Comment by sajadghawami** [comment]
AFAIK it is only supported for `webpack@5` and not `swc`
